lightly/loss/byol_loss.py

```diff
@@ -0,0 +1,189 @@
+
+from models.mlp import MLP
+
+
+class ExperimentParams:
+    """
+    Encapsulate hyper-parameters for performing and logging experiments.
+    """
+
+    def __init(self, use_momentum, beta, augment_functions, net, projection_size, projection_hidden_size, layer):
+        # TODO: implement momentum
+        self.use_momentum = use_momentum
+        self.beta = beta
+        self.augment_functions = augment_functions
+        self.net = net
+        self.projection_size = projection_size
+        self.projection_hidden_size = projection_hidden_size
+        self.layer = layer
+
+    def __str__(self):
+        return "Running experiment with the following hyper-parameters.\n" \
+               "Momentum: {}\nBeta: {}\nAugment Function: {}\nNet: {}\nProjection Size: {}\n" \
+               "Projection Hidden Size: {}\n" "Layer: {}".format(self.use_momentum, self.beta, self.augment_functions,
+                                                                 self.net, self.projection_size,
+                                                                 self.projection_hidden_size, self.layer)
+
+
+class NetWrapper(nn.Module):
+    """
+    Utilities for accessing hidden layers.
+    """
+    def __init__(self, net, projection_size, projection_hidden_size, layer=-2):
+        super().__init__()
+        self.net = net
+        self.layer = layer
+
+        self.projector = None
+        self.projection_size = projection_size
+        self.projection_hidden_size = projection_hidden_size
+
+        self.hidden = None
+        self.hook_registered = False
+
+    def _find_layer(self):
+        if type(self.layer) == str:
+            modules = dict([*self.net.named_modules()])
+            return modules.get(self.layer, None)
+        elif type(self.layer) == int:
+            children = [*self.net.children()]
+            return children[self.layer]
+        return None
+
+    def _hook(self, _, __, output):
+        self.hidden = self.flatten(output)
+
+    def _register_hook(self):
+        layer = self._find_layer()
+        assert layer, f'hidden layer ({self.layer}) not found'
+        layer.register_forward_hook(self._hook)
+        self.hook_registered = True
+
+    @singleton('projector')
+    def get_projector(self, hidden):
+        _, dim = hidden.shape
+        projector = MLP(dim, self.projection_size, self.projection_hidden_size)
+        return projector.to(hidden)
+
+    def get_representation(self, x):
+        if self.layer == -1:
+            return self.net(x)
+
+        if not self.hook_registered:
+            self._register_hook()
+
+        _ = self.net(x)
+        hidden = self.hidden
+        self.hidden = None
+        assert hidden is not None, f'hidden layer {self.layer} never emitted an output'
+        return hidden
+
+    @staticmethod
+    def default(val, def_val):
+        return val if val else def_val
+
+    @staticmethod
+    def flatten(t):
+        return t.reshape(t.shape[0], -1)
+
+
```

Comment:
The classes `OnlineModel` and `TargetModel` will receive an abstract parent class with #57 (I will make a PR soon).

IMO it would be best to reuse the parent class as soon as it's ready.